canonical/ds25
# feat: Generator CLI support

## Done

Creates an adapter interface that allows generators to easily prompt for input in either CLI or interactive mode.
This allows generators to call for both input types using a very small API:
```ts
  constructor(args: string | string[], options: Record<string, unknown>) {
    super(args, options);

    if (!isInInteractiveMode(this)) this.answers = getCliAnswers(this, this.questions))
  }

  async prompting() {
    if (!this.inInteractiveMode) return;
    this.answers = await getInteractiveAnswers(this, this.questions);
  }
```

This will allow the generator to be executed programmatically, enhancing its testability.

## QA

- `bun i`
- `yo @canonical/canonical-ds:component` - see that the generator now fails due to missing required positional argument
- `yo @canonical/canonical-ds:component path/to/button` - see that the generator creates a default button component
- `yo @canonical/canonical-ds:component path/to/button --includeStorybook --includeStyles` - see the generators also includes storybook and style files
- `yo @canonical/canonical-ds:component --interactive` - see that the generator still work in interactive mode
- `yo @canonical/canonical-ds:component --help` - see the generator displays input parameter help

### PR readiness check

  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.

## Screenshots
![Screenshot 2024-12-12 at 18 14 36](https://github.com/user-attachments/assets/8e204205-9ac8-4771-83df-59aedfe8f36c)
